### PR TITLE
Remove-DbaBackup, fixes #970

### LINKED ISSUE #970: Remove-DbaBackup, Slow Folder Removal and Failed Deletion of Files
### Is this a *feature* OR *bug*:
Bug

### System Details

- Operating system name and version:
- Output from `$PSVersionTable`:

```
Evaluate $PSVersionTable in PowerShell and paste the output here
```

- Output of dbatools version:

```
Evaluate (Get-Module dbatools -ListAvailable).Version and paste output here
```

- SQL Server version for source/target

### Steps to Reproduce

When the option to remove emp
foldersdeleted
ty folders is used it takes a long time to delete a small amount of folders.  See attached.

Also, when you delete a large amount of tlogs backups for example if the file is deleted prior to the process getting to that file it throws a red error.

### Attached Logs or any exception errors

- Use of Start-Transcript can help collection of console output and exceptions
- Certain commands will generate an exception log that you can find in following directory: `dir $env:USERPROFILE\Documents\dbatools-exceptions.txt`

### PR DESCRIPTION
**Pull request by niphlod**
Fixes #970 

Changes proposed in this pull request:
 - avoid long waits to build up file lists, everything is pipelined, so concurrent alteration of the underlying filestructure are less likely to impact the command itself
 - new logging subsystem
 - returns the removed files and directory fullnames for empty folder removals
- avoids top-down recursion when removing empty folders (which ultimately was the reason behind the relative slowness described)

tl;dr: on 302k files nested into 5k dirs on a local SSD the process passed from 20min to 6min

### CODE REVIEW

**Comment by niphlod** [comment]
@FriedrichWeinmann : I'd need a crash course on how to handle possible throws with the new system. stop-function is the solution there ?
